src/Pool.ts

```diff
@@ -339,32 +382,80 @@ export class Pool<T> {
+      const free = this.free.pop();
+
+      if (free === undefined || pendingAcquire === undefined) {
+        const errMessage = 'this.free was empty while trying to acquire resource';
+        this.log(`Tarn: ${errMessage}`, 'warn');
+        throw new Error(`Internal error, should never happen. ${errMessage}`);
       }
 
-      this.pendingAcquires.shift();
-      this.free.pop();
-      this.used.push(free.resolve());
+      // Make sure that pendingAcquire that is being validated is not lost and
+      // can be freed when pool is destroyed.
+      this.pendingValidations.push(pendingAcquire);
 
-      //At least one active resource, start reaping
-      this._startReaping();
+      // Must be added here pre-emptively to prevent logic that decides
+      // if new resources are created will keep on working correctly.
+      this.used.push(free);
 
-      pendingAcquire.resolve(free.resource);
-    }
+      // if acquire fails also pending validation, must be aborted so that pre reserved
+      // resource will be returned to free resources immediately
+      const abortAbleValidation = new PendingOperation<boolean>(this.acquireTimeoutMillis);
+      pendingAcquire.promise.catch(err => {
+        abortAbleValidation.abort();
+      });
 
-    // If we destroyed invalid resources, we may need to create new ones.
-    if (didDestroyResources) {
-      this._tryAcquireOrCreate();
+      abortAbleValidation.promise
+        .catch(err => {
+          // There's nothing we can do here but log the error. This would otherwise
+          // leak out as an unhandled exception.
+          this.log('Tarn: resource validator threw an exception ' + err.stack, 'warn');
+          return false;
+        })
+        .then(validationSuccess => {
+          try {
+            if (validationSuccess) {
+              // At least one active resource exist, start reaping.
+              this._startReaping();
+              pendingAcquire.resolve(free.resource);
+            } else {
+              remove(this.used, free);
+              this._destroy(free.resource);
+
+              // is aquire was canceled, failed or timed out already
+              // no need to return it to pending queries
+              if (!pendingAcquire.isRejected) {
+                this.pendingAcquires.unshift(pendingAcquire);
+              }
+
+              // Since we destroyed an invalid resource and were not able to fulfill
+              // all the pending acquires, we may need to create new ones or at
+              // least run this acquire loop again to verify it. But not immediately
+              // to prevent starving event loop.
+              setTimeout(() => {
+                this._tryAcquireOrCreate();
+              }, 0);
+            }
+          } finally {
+            remove(this.pendingValidations, pendingAcquire);
+          }
+        });
+
+      // try to validate
+      this._validateResource(free.resource)
```

Comment:
we can use async/await

Reply:
I didn't want to go there yet. I'd rather have that change done as a separate PR and change it completely. 

Actually this code is already written in typescript so we could have done that also earlier and compile it for node 6. Though I'm glad that we didn't since I actually had to do lots of running the code in debugger to figure out why stuff didn't work and to figure out how it should work. Code didn't seem to have source maps set properly, so this way it was lot easier to follow up the execution from compiled code.